Martin fuck up

furadnik opened the pull request.

MARTESING committed:
Added tests and second variant of the model.

I added variant of the model where ownership is fixed from the start and also a "decay" of votes for non-owners of a resource, named other_loss.
I also used the constants as parameters for the game, so they can be easily changed.
Constants PLAYERS and RESOURCES remain, I do not know if we want to change them in tests?
I further added some trivial test for the utility function and for the game, but there should be more for the game, I think.
I added tests for checking the pure strategies and they show that we do not always have pure strategies.
File data_large.csv gives data not for the regions, but for the towns. I wanted to check if we are able to apply the model to this number if we had only pure strategies, but it seemed, not surprisingly, to be super slow.

MARTESING committed:
Making test more general with general parameter.

Something I forgot to change in the last commit after testing.

furadnik commented on senator/__main__.py:
```diff
@@ -1,8 +1,15 @@
 
-game.run_greedy_search()
\ No newline at end of file
+votes, initial_utility, is_owner = parse_data(load_data(FILE_PATH))
+game = Game(initial_utility, votes, is_owner, OWNER_LOSS, OTHER_LOSS, OWNER_TRASHOLD)
+#game = Game(np.zeros((3,17)),votes)
+
+actions_in_steps, final_utility, strategies = game.run_greedy_search()
+print(strategies)
+#print(final_utility.sum(axis=1))
```
Veškerý kód který se volá, včetně tohohle, se většinou dává do nějaké funkce, např. `main`, a pak na konci tohohle souboru dej
 ```python
if __name__ == '__main__':
    main()
```
To je taková magická formule, která spustí funkci `main` pokud je tenhle soubor spuštěn přímo z command liny. Můžeš se zeptat chatu na to co to přesněji dělá, myslím že by to měl být schopen docela dobře vysvětlit.

MARTESING replied:
Ok, done

furadnik commented on senator/game.py:
```diff
@@ -63,7 +64,7 @@ def rm_plus_step(self) -> None:
         self.iterations += 1
         strategy = self.current_strategy
-        player = self.iterations % self._num_players
+        player = (self.iterations -1) % self._num_players
```
Možná by bylo přehlednější místo té `-1` prostě dát `self.iterations += 1` na konec téhle funkce. To by nemělo nic rozbít ne?

MARTESING replied:
Yes, přehodil jsem.

furadnik commented on senator/game.py:
```diff
@@ -84,15 +85,21 @@ def nash_conv(self) -> float:
-    def __init__(self, initial_utility: np.ndarray, votes: np.ndarray):
-        assert len(initial_utility.shape) == 2, "Initial utility should be a 2D array"
+    def __init__(self, initial_utility: np.ndarray, votes: np.ndarray, is_owner: np.ndarray, 
+    owner_loss = 0.8, other_loss = 0, owner_trashold: float = 2.0):
```
Tady ti chybí u některých parametrů typování. Also, funkce `__init__` se typuje jako že vrací `None`, tj. `def __init__(...) -> None:`

furadnik commented on senator/game.py:
```diff
@@ -102,21 +109,29 @@ def solve_node_via_rm_plus(self, node: Node, iterations: int = 1000) -> None:
             print(f"Nash Convexity is {conv}, something is wrong...")
 
-    def run_greedy_search(self, num_expansions: int = 10):
-        """
-        Run a greedy search algorithm, periodically expanding the tree. Each expansion is done by first solving the
-        node using the RM+ algorithm, sampling a joing action, and adding the child to the tree.
-        TODO: decide how to do the expanding. Baseline is to just do a path of length 5.
-        """
```
Tomuhle se říká docstring, a měl by dokumentovat co funkce dělá, co vrací, co znamenají jaké její parametry atd. Neříkám že by to muselo být u každé funkce, ale většinou je dobrej nápad ho psát. Až to budeš číst za týden tak se ti to bude hodit :D

furadnik replied:
PS docstring můžeš napsat i k třídě, třeba 
```python
class Foo:
    """This is class Foo, it does foo stuff."""
```

MARTESING replied:
Průběžně doplním.

furadnik commented on senator/game.py:
```diff
@@ -102,21 +109,29 @@ def solve_node_via_rm_plus(self, node: Node, iterations: int = 1000) -> None:
+
+        print(f"Initial utility\nu: {utility.sum(axis=1)}")
+
+        for i in range(num_expansions):
```
Tady děláš `for i in ...`, ale pak to `i` nikde nepoužíváš (pokud nejsem slepý). V takovém případě je lepší udělat `for _ in ...`, čímž dáváš na jevo že skutečně to `i` nebudeš potřebovat.

MARTESING replied:
Yes, tohle je nějaký zmetek po předchozí iteraci kódu, kde jsem přes ty i-čka vstupoval do pole. Pak jsem přišel na append, a tohle už neupravil. Upraveno.

MARTESING replied:
Počkat, já tam dole printím číslo dne, kde to i používám. Takže nechávám i.

furadnik replied:
Jo sorry

furadnik commented on senator/utility.py:
```diff
@@ -1,21 +1,29 @@
+                is_owner: np.ndarray = np.zeros((PLAYERS, RESOURCES)),
+                owner_loss: float = 0.8, 
+                other_loss: float = 0,
+                owner_trashold: float = 2.0) -> np.ndarray: 
```
Tohle mi přijde ok, akorát teda možná by mi přišlo přirozenější dát tam něco jako `-1` jako tu "vypínací hodnotu". Ale jak chceš, za mě skvělé

MARTESING replied:
Dal jsem -1

furadnik commented on senator/utility.py:
```diff
@@ -1,21 +1,29 @@
+    if owner_trashold <= 1.0:
+        is_owner = previous_utility >= owner_trashold * votes #original variant of the model, where is_owner is computed by using a trashold
+       
+    current_utility = ((previous_utility * owner_loss) * is_owner) + (previous_utility * other_loss) * (1- is_owner)
```
Jedna hodně hnidopišská, hodně se ti rozchází kdy děláš mezery kolem čárek, kolem znamének atd. Existujou extensions do VS code, který dělají automatický formátování. Doporučuju si něco najít a nainstalovat a pak to prostě neřešit, protože se to vyřeší za tebe :D vím že asi `pylint` má nějaký formátování, pak je `black` (ten je hodně striktní), `autopep8`. Zkus se po tom kouknout, nebo @juliakrizanova jestli třeba něco nemá, ta asi používá VS code? Nebo případně Richard

furadnik commented on senator/utility.py:
```diff
@@ -42,6 +54,7 @@ def load_data(file_path: str) -> np.ndarray:
 
 def parse_data(data: np.ndarray) -> tuple[np.ndarray, np.ndarray]: 
```
Špatně type hints :) To mi připomíná, další rozšíření je `mypy`, který dělá automatický checking typů a upozorní tě když děláš něco divnýho. @juliakrizanova máš s tím zkušenosti ve VS code?

MARTESING replied:
To tam nejspíš také zůstalo z nějaké předchozí iterace. Počkám, co poradí mamka =D

juliakrizanova replied:
S tymto skusenosti nemam, ale znie to ako uzitocna vec. Pozriem ako to spojazdnit vo VScode a dam vediet:D

furadnik commented on senator/__main__.py:
```diff
@@ -1,8 +1,15 @@
 from senator.game import *
 
-votes, initial_utility = parse_data(load_data(FILE_PATH))
+OWNER_LOSS = 0.59
+OTHER_LOSS = 0.19
+OWNER_TRASHOLD = 0.18 #Set 2.0 to change the model TODO: find better solution than 2.0
```
Tyhle věci by nemusely být definované jako konstanty, ale nastavené jako parametry přes `argparse` https://docs.python.org/3/library/argparse.html. Zeptej se na to chatu.

Point je že pak můžeš dělat věci jako `python -m senator --owner_trashold=2`, s tím že když tam nedáš nic tak tam bude nějaký default

MARTESING committed:
I fixed everything

furadnik commented on senator/__main__.py:
```diff
@@ -1,15 +1,49 @@
+    parser.add_argument(
+        "--owner_loss",
+        type=float,
+        default=0.8,
+        help="Loss for the owner (default: 0.5)",
```
Ty help messages jsou k tomu aby když dáš `python -m senator --help` se vyprintily všechny možný optiony a informace o nich. Ty tam teď dáváš i defaultní hodnoty. To je ok, ale třeba tady jsi očividně ten default změnil, a pak neupdatoval ten help.

Aby se tohle nestávalo, můžeš tomu argparseru říct ať to dělá sám. To se dělá tak, že mu nastavíš help formatter, jinými slovy tu definici `parser`u změníš na 
```python
    parser = argparse.ArgumentParser(
        description="Run the senator game simulation with specified parameters.",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
```

MARTESING replied:
HOHOHO! Done

furadnik reviewed (commented):
Je to pěkné. Jen jeden detail ohledně argparseru, jinak za mě skvělé

furadnik commented on senator/__main__.py:
```diff
@@ -1,8 +1,49 @@
+    )
+    # game = Game(np.zeros((3,17)),votes)
+
+    actions_in_steps, final_utility, strategies = game.run_greedy_search()
```
Ještě bych přidal argument do argparseru na kolik kroků se to počítá (což je parametr `run_greedy_search`)

MARTESING replied:
Ok, je tam argument num_iterations, který se tam dává.

furadnik replied:
Jo počkat, tak to v tom případě nedávej do té funkce jako parametr, ale vevnitř používej to `self.num_iterations` který tam už máš uložený z `__init__`u.

MARTESING replied:
Zněnil jsem na num_steps a upravil

furadnik replied:
:+1:

furadnik commented on senator/utility.py:
```diff
@@ -25,13 +33,17 @@ def get_utility(previous_utility: np.ndarray, chosen_resources: list[int],
 
-def get_utility_matrix(previous_utility: np.ndarray, votes: np.ndarray) -> np.ndarray:
+def get_utility_matrix(previous_utility: np.ndarray, votes: np.ndarray, 
+                       is_owner: np.ndarray,
+                         owner_loss: float = 0.8, other_loss: float = 0.0, owner_trashold: float = -1) -> np.ndarray:
```
Tady tenhle file by chtěl ještě prohnat tím blackem :)

furadnik commented on senator/__main__.py:
```diff
@@ -1,8 +1,49 @@
+
+    args = parser.parse_args()
+
+    votes, initial_utility, is_owner = parse_data(load_data(FILE_PATH))
```
ještě jedna věc mě napadla že by mohla být fajn. místo `FILE_PATH` přidat argument `file_path`, a `FILE_PATH` mu nastavit jako default.

furadnik replied:
Já myslel jako argument do toho `parser`u ať to jde nastavit z command line :D

MARTESING replied:
ok, upravím

MARTESING committed:
Minor changes based on Filip's feedback

MARTESING committed:
Fixed wrong parameter name

MARTESING committed:
added experiments modul with parameter_search and plotting

MARTESING committed:
Accidently deleted files with data

MARTESING committed:
fixed what papa wants

MARTESING committed:
hopefully what the boss wants =D

MARTESING committed:
typo and change from current to average strategy

DavidSych commented on senator/game.py:
```diff
@@ -11,22 +11,33 @@ def regret_matching(regret: np.ndarray) -> np.ndarray:
+    def average_strategy(self) -> np.ndarray:
+        strategy = regret_matching(self._cumulative_strategy)
+
+        strategy_sum = strategy.sum(axis=1)
```
Tohle není nikdy použitý. Navíc to není třeba, `regret_matching` už zajistí to, že je `strategy` správně normalizovaná. Takže můžeš prostě vrátit strategy a čus

MARTESING replied:
To přesně dělá current_strategy. Na co jsem měl teda udělat ještě average?

MARTESING replied:
Ok, už jsme se pochopili

DavidSych commented on senator/game.py:
```diff
@@ -84,16 +105,45 @@ def nash_conv(self) -> float:
+        initial_utility: np.ndarray,
+        votes: np.ndarray,
+        is_owner: np.ndarray,
+        num_iterations: int = 7,
```
Souhlasim s Otrokem, že `iterations` už je použité pro to řešení pomocí regret matchingu (i když ta fce se jmenuje `rm_step`...), takže to chce nějak sjednotit a použít někde step a jinde konzistentně iteration.

MARTESING replied:
Ok, dal jsem num_step

MARTESING commented on senator/game.py:
```diff
@@ -11,20 +11,23 @@ def regret_matching(regret: np.ndarray) -> np.ndarray:
         self._cumulative_regret = np.zeros((self._num_players, self._num_resources))
-        self._cumulative_positive_regret = np.zeros((self._num_players, self._num_resources))
-        self.children = {}  # key: action, value: Node
+        self._cumulative_positive_regret = np.zeros(
```
ok

MARTESING commented on senator/game.py:
```diff
@@ -38,37 +41,45 @@ def sample_joint_action(self) -> np.ndarray:
 
         self._cumulative_positive_regret[player] = np.maximum(
-                                self._cumulative_positive_regret[player] + instantaneous_regret, 0)
+            self._cumulative_positive_regret[player] + instantaneous_regret, 0
```
ok

MARTESING commented on senator/game.py:
```diff
@@ -38,37 +41,45 @@ def sample_joint_action(self) -> np.ndarray:
+        )
 
     def nash_conv(self) -> float:
         """
```
ok

MARTESING committed:
Last upadates according to Filip and David

MARTESING committed:
Last updates according to Filip and David

MARTESING committed:
dropping unnecessary variable

MARTESING committed:
...

furadnik commented on senator/__main__.py:
```diff
@@ -1,8 +1,62 @@
+    parser.add_argument(
+        "--file_path",
+        type=float,
+        default=-1,
```
What about
```python
default=FILE_PATH
```
?

furadnik replied:
Also, this needs to be passed to `load_data` below.

MARTESING replied:
ok, done

MARTESING committed:
fixed nash_conv and args

MARTESING committed:
fixing argparse and nash conv

MARTESING commented:
Tak pánové, můžete znovu zhejtit. Potom zavřeme a vrhnu se na ty intenzity

DavidSych commented on senator/game.py:
```diff
@@ -38,37 +47,47 @@ def sample_joint_action(self) -> np.ndarray:
+        ):  # We start from the last opponent because `np.tensordot` collapses axes
+            utility_matrix = np.tensordot(
+                utility_matrix, strategy[opponent], axes=([opponent], [0])
+            )  # TODO: check if this is correct
```
Jestli jsi to checknul, můžeme tohle smazat :)

MARTESING replied:
S Filipem jsme to nějak řešili hned ten první den, co jsi nám to posílal, myslím, že ok.

furadnik replied:
Jo tohle TODO smazat

DavidSych commented on senator/game.py:
```diff
@@ -78,45 +97,93 @@ def nash_conv(self) -> float:
+            parent=node, utility_matrix=child_utility_matrix
+        )
+
+    def solve_node_via_rm_plus(self, node: Node, iterations: int = 2000) -> None:
```
Těch 2000 iterací je možná trochu random. Třeba některé utility matice budou potřebovat mín a některé víc. Možná by se tady dalo udělat něco jako 
`while node.nash_conv > 0 and iterations < 1e4:`
a tam házer rm_plus_step?

MARTESING replied:
Já jsem večer pustil experimenty na 10 000 iterací, protože pořád něco padalo a stejně se našla nastavení parametrů, kde Nash Conv padl... 

Takhle to nakombinovat dobrý nápad.

MARTESING replied:
Ikdyž teda našel jsem i kombinaci parametrů, kde to neseběhlo za 10 000 iterací ani pro rm_plus_step, ale je jich fakt něco jako 1-3%.

DavidSych commented on senator/game.py:
```diff
@@ -78,45 +97,93 @@ def nash_conv(self) -> float:
+
+            utility = get_utility(
+                utility,
+                joint_action,
```
Tady mi to hlásí typovací chybu, protože `current_node.sample_joint_action()` vrací `np.array`, ale `get_utility` čeká tohle jako `list[int]`.

MARTESING replied:
Přehodím na np.array

furadnik commented on senator/__main__.py:
```diff
@@ -1,8 +1,62 @@
+        "--owner_trashold",
+        type=float,
+        default=-1,
+        help="Threshold for the owner. Set to -1 to change the model.",
```
Change the model to what? A butterfly? A unicorn? Another model?

MARTESING replied:
Unicorn would be nice!

Fair point, will change that.

furadnik commented on senator/__main__.py:
```diff
@@ -1,8 +1,62 @@
 from senator.utility import *
 from senator.game import *
```
Ještě detail: `import *` se skoro nepoužívá, protože pak není vidět co importuješ.... Lepsí je buď udělat `import senator.utility` a pak mít třeba `senator.utility.load_data`, a nebo explicitně napsat `from senator.utility import load_data`.

furadnik commented on senator/game.py:
```diff
@@ -78,45 +97,93 @@ def nash_conv(self) -> float:
+        for _ in range(iterations):
+            node.rm_step()
+
+        if (conv := node.nash_conv()) >= 0.001:
+            print(f"Nash Convexity is {conv}, something is wrong...")
```
Možná by stálo za to udělat z tohohle svou vlastní metodu `self.check_nash_conv()`?

furadnik replied:
Ale jestli se ti do toho nechce tak tohle resolvni a kašli na to xD

furadnik commented on experiments/parameter_search.py:
```diff
@@ -0,0 +1,87 @@
+    )
+
+    # cycle accross parameters
+    for i in range(num_owner_iterations + 1):  # TODO: no for?
```
Ten TODO note možná odstranit :)

furadnik commented on senator/game.py:
```diff
@@ -78,45 +97,93 @@ def nash_conv(self) -> float:
+            self.owner_loss,
+            self.other_loss,
+            self.owner_trashold,
+        )  # TODO: see what it takes to get the utility matrix
```
Tenhle TODO nevim co znamená. Taky smazat?

furadnik commented on senator/game.py:
```diff
@@ -11,20 +11,29 @@ def regret_matching(regret: np.ndarray) -> np.ndarray:
     @property
     def current_strategy(self) -> np.ndarray:
-        strategy = regret_matching(self._cumulative_positive_regret)
+        strategy = regret_matching(self._cumulative_regret)
+        return strategy
+
+    @property
+    def average_strategy(self) -> np.ndarray:
+        strategy = regret_matching(self._cumulative_strategy)
+
```
Tady je podezřelý že máš dvě funkce který vypadají skoro stejně, a jedna z nějakého důvodu magicky dává na místo regretu tu cum strategy.

Zaprvé, @DavidSych je to takhle ok?

Zadruhé, jestli jo, tak bych tu funkci `regret_matching` přejmenoval na něco jako `normalize_to_strategies` nebo něco v tom smyslu aby bylo jasný co dělá, a proč teda je ok dávat tam něco co není regret.

furadnik replied:
Jo ok tak v tom případě bych tu funkci co to volá jen přejmenoval na něco jako `normalize....`. :)

furadnik commented on senator/game.py:
```diff
@@ -78,45 +97,93 @@ def nash_conv(self) -> float:
+            self.other_loss,
+            self.owner_trashold,
+        )  # TODO: see what it takes to get the utility matrix
+        node.children[tuple(joint_action)] = Node(
```
Před tenhle řádek bych dal 
```python
assert tuple(joint_action) not in node.children.keys()
```
protože předpokládám že nechceme přidávat node child když už nějaký existuje. Nebo? Co ty na to @DavidSych ?

DavidSych replied:
Fakt tady chceme `assert`? Nevim jak tohle navazuje na zbytek kodu, ale určitě nechceme `Node` premazat. Takže možná jen 
```
if tuple(joint_action) not in node.children.keys():
     node.children[tuple(joint_action)] = Node(...)
```

furadnik replied:
Já myslím že jakmile voláme `add_child` na něco co už ten child má tak je to docela fuckup ne?

furadnik replied:
Ale je to synkův kód, tak ať se s tím děje vůle ne boží, ale jeho

DavidSych replied:
Asi souhlas no :D Já si představoval, že ty děti se budou přidávat pro náhodně vygenerované akce z equilibiálních strategií. Takže se může stát, že nasampluješ tu joint akci 2x, ale můžeš ten if hodit i mimo tuhle fci.

furadnik replied:
Jasný, jako může být že ten if bude tady, asi to dává smysl. Ale pak bych tenhle fakt hodil do docstringu té metody.

furadnik reviewed (commented):
Už převážně fakt kosmetický hlouposti. Pár otázek mám i k tomu regretu, ale nerozumim tomu takže když David řekne že jsou to blbosti tak mi to klidně zavřete. Jinak je to pěkný. Kdybys to mistře chtěl už zamergovat tak na ty moje poznámky asi neber moc ohled :)

furadnik commented:
Jo je otázka jestli chceme mít v tom repu takhle experiment results... Asi bych to dával aspoň do svýho adresáře `results/`.

MARTESING committed:
added intensities and minor fixes

MARTESING committed:
tests fix

MARTESING committed:
Create script.py

script for David

MARTESING committed:
Add files via upload

MARTESING committed:
Add files via upload

MARTESING committed:
Update script.py

typos

MARTESING committed:
Update script.py

script for David

MARTESING committed:
Update script.py

Rozdělení parametrů LAST_WEEK_LENGTH A WEEK_LENGTH, aby mohl případně senátor dodávat data i po obdobích nepravidelné délky.

MARTESING committed:
Update game.py

Upravené na moji verzi, kde je i ta funkce compute_modified_utility.

furadnik committed:
Add matplotlib and pandas to requirements

furadnik committed:
support leaving default actions from prev week

furadnik committed:
ANO guy utilities edit

furadnik committed:
add .venv ignore

furadnik committed:
tyden 0-2

furadnik committed:
add comment about this_week_actions not being none
